src/Language/C/Monad.hs

```diff
@@ -415,7 +415,10 @@ liftSharedLocals prog = do
     let (globs, shared) = unzip $ map (extractDecls (`Set.member` uvs)) oldglobs
         sharedList = Set.toList $ Set.unions shared
         sharedDecls = map (\ig -> C.DecDef ig (SrcLoc NoLoc)) sharedList
-    void $ globals <<.= (globs ++ sharedDecls)
+    -- Reverse is a trick that ensures the correct order of declarations for arrays
+    -- and their wrapper pointers. It depends on the naming schema of identifiers:
+    -- arrays are prefixed with underscores, while their wrappers are not.
+    void $ globals <<.= (globs ++ reverse sharedDecls)
```

Comment:
I am not very satisfied with this trick but a more precise solution leads very far: dependencies between variables. Maybe it could be implemented as a custom ordering for the set collecting the used local variables, but now I wanted to solve another problem, that directly affects C compilation of the generated code.